internal/core/entity/entity_test.go

```diff
@@ -0,0 +1,106 @@
+package entity
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestUpstream(t *testing.T) {
+	upstreamStr := `{"type":"roundrobin"}`
+
+	var upstream Upstream
+	err := Unmarshal([]byte(upstreamStr), &upstream)
+	assert.Nil(t, err)
+
+	jsonStr, err := Marshal(&upstream)
+	assert.Nil(t, err)
+	assert.Equal(t, upstreamStr, string(jsonStr), "Test unnamed attribute")
+
+	upstream.ServiceName = "TEST-SERVICE"
+	upstream.DiscoveryType = "nacos"
+
+	embedElm(reflect.ValueOf(&upstream), upstream.All)
+	expectedAll := map[string]interface{}{
+		"discovery_type": "nacos",
+		"service_name":   "TEST-SERVICE",
+		"type":           "roundrobin",
+	}
+	assert.True(t, reflect.DeepEqual(upstream.All, expectedAll), "Test latest attribute")
+
+	upstream.DiscoveryArgs = &UpstreamArg{}
+	upstream.DiscoveryArgs.NamespaceID = "TEST-NAMESPACE"
+	upstream.DiscoveryArgs.GroupName = "TEST-GROUP"
+
+	embedElm(reflect.ValueOf(&upstream), upstream.All)
+	expectedArg := map[string]interface{}{
+		"namespace_id": "TEST-NAMESPACE",
+		"group_name":   "TEST-GROUP",
+	}
+	assert.True(t, reflect.DeepEqual(upstream.All["discovery_args"], expectedArg), "Test latest attribute")
+}
+
+func TestRoute(t *testing.T) {
+	routeStr := `{"uris":["/*"]}`
+
+	var route Route
+	err := Unmarshal([]byte(routeStr), &route)
+	assert.Nil(t, err)
+
+	jsonStr, err := Marshal(&route)
+	assert.Nil(t, err)
+	assert.Equal(t, routeStr, string(jsonStr), "Test unnamed attribute")
+
+	route.Upstream = &UpstreamDef{}
+	route.Upstream.ServiceName = "TEST-SERVICE"
+	route.Upstream.DiscoveryType = "nacos"
+
+	nodes := []*Node{
+		{Host: "test.com"},
+	}
+	route.SetNodes(nodes)
+
+	embedElm(reflect.ValueOf(&route), route.All)
+	expectedAll := map[string]interface{}{
+		"uris": `[/*]`,
+		"upstream": map[string]interface{}{
+			"discovery_type": "nacos",
+			"service_name":   "TEST-SERVICE",
+			"nodes":          nodes,
+		},
+	}
+	assert.True(t, reflect.DeepEqual(route.All["upstream"], expectedAll["upstream"]), "Test latest attribute")
```

Comment:
Should we compare route.All with expectedAll?

Reply:
I have made the changes.